packages/flutter/lib/src/material/carousel.dart

```diff
@@ -0,0 +1,751 @@
+  final double? shrinkExtent;
+
+  /// Whether the carousel should keep scrolling to the next/previous items to
+  /// maintain the original layout.
+  ///
+  /// Defaults to false.
+  final bool itemSnapping;
+
+  /// An object that can be used to control the position to which this scroll
+  /// view is scrolled.
+  final CarouselController? controller;
+
+  /// The [Axis] along which the scroll view's offset increases with each item.
+  ///
+  /// Defaults to [Axis.horizontal].
+  final Axis scrollDirection;
+
+  /// Whether the carousel list scrolls in the reading direction.
+  ///
+  /// For example, if the reading direction is left-to-right and
+  /// [scrollDirection] is [Axis.horizontal], then the carousel scrolls from
+  /// left to right when [reverse] is false and from right to left when
+  /// [reverse] is true.
+  ///
+  /// Similarly, if [scrollDirection] is [Axis.vertical], then the carousel view
+  /// scrolls from top to bottom when [reverse] is false and from bottom to top
+  /// when [reverse] is true.
+  ///
+  /// Defaults to false.
+  final bool reverse;
+
+  /// Called when one of the [children] is tapped.
+  final ValueChanged<int>? onTap;
+
+  /// The extent the children are forced to have in the main axis.
+  ///
+  /// The item extent should not exceed the available space that the carousel
+  /// occupies to ensure at least one item is fully visible.
+  ///
+  /// This must be non-null.
+  final double itemExtent;
+
+  /// The child widgets for the carousel.
+  final List<Widget> children;
+
+  @override
+  State<CarouselView> createState() => _CarouselViewState();
+}
+
+class _CarouselViewState extends State<CarouselView> {
+  late double _itemExtent;
+  CarouselController? _internalController;
+  CarouselController get _controller => widget.controller ?? _internalController!;
+
+  @override
+  void initState() {
+    super.initState();
+    if (widget.controller == null) {
+      _internalController = CarouselController();
+    }
+    _controller._attach(this);
+  }
+
+  @override
+  void didChangeDependencies() {
+    super.didChangeDependencies();
+    _itemExtent = widget.itemExtent;
+  }
+
+  @override
+  void didUpdateWidget(covariant CarouselView oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    if (widget.controller != oldWidget.controller) {
+      oldWidget.controller?._detach(this);
+      if (widget.controller != null) {
+        _internalController?._detach(this);
+        _internalController = null;
+        widget.controller?._attach(this);
+      } else { // widget.controller == null && oldWidget.controller != null
+        _internalController = CarouselController();
+        _controller._attach(this);
+      }
+    }
+    if (widget.itemExtent != oldWidget.itemExtent) {
+      _itemExtent = widget.itemExtent;
+    }
+  }
+
+  @override
+  void dispose() {
+    _controller._detach(this);
+    if (widget.controller == null) {
+      _controller.dispose();
+    }
```

Comment:
I think it would be clearer and less error-prone to do:

```suggestion
    _internalController?.dispose();
```